numba/core/interpreter.py

```diff
@@ -955,6 +956,277 @@ def peep_hole_delete_with_exit(func_ir):
     return func_ir
 
 
+def peep_hole_fuse_dict_add_updates(func_ir):
+    """
+    This rewrite removes d1.update(d2) calls that
+    are between two dictionaries, d1 and d2 in the
+    same basic block. This pattern can appear as a
+    result of a Python 3.10 bytecode changes, which
+    prevents large constant literal dictionaries with
+    > 15 elements from being constant. If both dictionaries
+    are constant dictionaries defined in that block and
+    neither is used between the update call, then we replace d1
+    with a new definition that combines the two dicitonaries.
+
+    Python 3.10 may also rewrite an individual dictionary as an empty
+    build_map + many map_add, so those expressions are also replaced
+    with a constant build map.
+
+    When running this algorithm we do not remove any maps, even if they
+    were fused into an update to protect against strange choices in user
+    code. Here we depend on dead code elimination to run later
+    and remove any unnecessary maps.
```

Comment:
Yes. Since the implementation only looks at particular basic block, the implementation does not/cannot determine that d2 is unused in the entire IR. This change seems like it should be handled with some dead code elimination.

There is nothing in this code the guarantees DCE runs. I didn't run DCE yet in case anything relied on explicit pattern matching and I assumed DCE would run later in the pipeline. I'm guessing this is an erroneous assumption and causes a unused dictionary to be allocated consisting of all of d2's entries.